Implement an additional check in the reducer for setting the debtee to avoid information loss

thomaseizinger opened the pull request:
Fixes #13.

Initially, we relied on a property that no longer exists on the action. (`isNewMember`).
Therefore, this was always false and the debtee (if added via the input box), was never explicitly added.

By adding a check to the reducer function, we add new persons explicitly to the bill through a "PersonValue".

duffleit commented on src/routes/App/routes/new/modules/addBillForm/reducer.js:
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
 
-      if (debtee === state.debtee) {
-        return state;
+      if (isTransientMember(debtee)) {
```
calling a method and writing a comment about the return value of this method seems like a code smell IMHO. Maybe just rename the function to `isMemeberInCurrentBill`?

thomaseizinger replied:
Well, `isMemberInCurrentBill` is not completely correct. I explicitly want to check if the member is a __transient__ member, e.g. if they were added on the add-bill screen.

duffleit commented on src/routes/App/routes/new/modules/addBillForm/reducer.js:
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
+      let {debtee} = action.payload;
 
-      if (!debtee && !isNewMember) {
+      if (!debtee || debtee === state.debtee) {
```
how about extracting variables? Something like `debteeNotSet` and `debteeNotChanged`

thomaseizinger replied:
Good point, will do.

duffleit commented on src/routes/App/routes/new/modules/addBillForm/reducer.js:
```diff
@@ -29,29 +42,39 @@ export const addBillFormReducer = (state = initialState, action) => {
+        personValues: [
+          ...state.personValues,
+          createPersonValue(debtee)
+        ]
```
I would aggregate the personvalues in a single function and not within in the return statement, as it is duplicated in the `NewMemberAdded` reducer

thomaseizinger replied:
I was thinking about this. However, isn't it more readable to do all the construction of the new state in one place in comparison to have it scattered over several functions?

duffleit commented on src/routes/App/routes/new/modules/addBillForm/reducer.js:
```diff
@@ -1,23 +1,36 @@
 
-    return stateCopy
-  };
+    let isTransientMemberFn = isTransientMemberFactory(pseudoGlobalState);
```
is it good practice to call selectors in the reducer? I thought the way to do it is `state -> reducer -> selector ` and not `state -> selector -> reducer`. The introduction of a `pseudoGlobalState` just to use a selector seems to be a bad tradeoff for me. Pretty hard to understand.

duffleit replied:
In addition, I think the comment is the best indicator that this section is hard to understand by someone who has to read this code :)

thomaseizinger replied:
Well, depends on what you want to favor. I wanted to reuse the logic implemented in the selectors to check for transient members. IMO staying DRY is worth the introduction of the pseudo-state. After all, selectors are just functios.

duffleit commented on src/routes/App/routes/new/modules/addBillForm/selectors.test.js:
```diff
@@ -0,0 +1,51 @@
+  expect(isExistingMember("Thomas")).toBeFalsy();
+  expect(isExistingMember("David")).toBeFalsy();
+
+});
```
👍 🥇

thomaseizinger commented:
I have removed the complicated selector again and instead implemented a simple function inside the reducer. I also added some tests for the reducers and extracted action creators for the two actions I modified.

Check the commit messages for details :)

I have decided to keep the "duplication" of creating the PersonValues array in the NEW_MEMBER_ADDED and NEW_DEBTEE_SELECTED reducer because I think it is easier to read if all the spreading and overwriting of the previous state is in one place.
